### PR TITLE
Clarify reduction behavior for an empty set of values

### PR DESCRIPTION
**Pull request by gramalingam**
Clarify the behavior of the reduction-ops when reducing an empty set of values, by updating the test-cases and documentation.

It is useful in various edge-cases. For example, ReduceProd should return 1 for an empty tensor, and ReduceSum should return 0 for an empty tensor.

(See https://github.com/onnx/onnx/issues/3651#issuecomment-1454338883)

### Summary

ReduceSum ({}) = 0
ReduceProd ({}) = 1
ReduceMin ({}) = Max. value of datatype
ReduceMax ({}) = Min. value of datatype
ReduceLogSum ({}) = minus infinity or undefined for datatypes without minus infinity
ReduceLogSumExp ({}) = minus infinity or undefined for datatypes without minus infinity
ReduceMean ({}) = Undefined

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
>       res = np.minimum.reduce(data, axis=axes, keepdims=keepdims)
E       ValueError: zero-size array to reduction operation minimum which has no identity

etc. 

Need to create a special case?

**Comment by gramalingam** [comment]
> > ```
> >   res = np.minimum.reduce(data, axis=axes, keepdims=keepdims)
> > ```
> 
> E ValueError: zero-size array to reduction operation minimum which has no identity
> 
> etc.
> 
> Need to create a special case?

If we agree on the spec update, yes. We can discuss in today's meeting too.

File: onnx/backend/test/case/node/reducesum.py [hunk 1]
```diff
@@ -194,3 +194,27 @@ def export_empty_axes_input_noop() -> None:
             outputs=[reduced],
             name="test_reduce_sum_negative_axes_keepdims_random",
         )
+
+    @staticmethod
+    def export_empty_set() -> None:
```
Comment (xadupre) [inline onnx/backend/test/case/node/reducesum.py:199]: Maybe that would be worth adding example like `ReduceSum([[], []], axis=0)` and `ReduceSum([[], []], axis=1)`.

Comment (gramalingam) [reply onnx/backend/test/case/node/reducesum.py:199]: Done

**Comment by yuanyao-nv** [comment]
@gramalingam What should be the correct result for input shapes such as `{0, 5}`? Such inputs are still empty but would it be more consistent to return output with shape `{0, 1}` instead of a scalar value?

**Comment by gramalingam** [comment]
> @gramalingam What should be the correct result for input shapes such as `{0, 5}`? Such inputs are still empty but would it be more consistent to return output with shape `{0, 1}` instead of a scalar value?

If the input shape is 0x5, and if we reduce along axis=0, the output will have shape 1x5. This PR just clarifies what the value will be in this tensor. (The shape computation is just as before, no change to that, sorry I didn't make this clear.)

**Comment by yuanyao-nv** [comment]
> > @gramalingam What should be the correct result for input shapes such as `{0, 5}`? Such inputs are still empty but would it be more consistent to return output with shape `{0, 1}` instead of a scalar value?
> 
> If the input shape is 0x5, and if we reduce along axis=0, the output will have shape 1x5. This PR just clarifies what the value will be in this tensor. (The shape computation is just as before, no change to that, sorry I didn't make this clear.)
> 
> If the input shape is 0x5, and if we reduce along axis=1, the output will have shape 0x1. In this case, there is no change.
> 
> In short, this PR is about what happens when the reduced-axes have zero-size, it is not about what happens when the non-reduced-axes have zero-size.

Thanks for the explanation Rama.

Another question is regarding the default values for logsum, logsumexp, and mean. 
I'm not sure what's the best way to reason about it but I'm thinking about it like finding the identity element of an operation: suppose A, B are 1D arrays for simplicity, then `Reduce(Reduce(A), Reduce(B))` should be just `Reduce(Concat(A,B))`, which is `Reduce(B)` if A is empty.
For, sum, prod, min, max, your default values listed above agree with this formula.
For logsumexp, -infinity maintains invariance for dtypes with -infinity, and undefined otherwise, as you listed above.
However, for logsum, I think 0 should be the identity element.
Mean is a bit of a special operation and doesn't quite fit the formula above. But I think 0 would also make the most sense as the identity element?

Do you agree?

**Comment by gramalingam** [comment]
> However, for logsum, I think 0 should be the identity element.
> Mean is a bit of a special operation and doesn't quite fit the formula above. But I think 0 would also make the most sense as the identity element?

Yes, I agree with your basic intuition. However, LogSum is just like LogSumExp: it is the Log of a reduced value, and, hence, it is log(0). 

As for Mean, it is not a direct reduction op. It is the composition of two reductions: namely, ReduceSum(...)/Count(...) ... which becomes zero-divided-by-zero. Best to treat it as undefined (though we could use NaN for floating-values).

**Comment by gramalingam** [comment]
> Do you agree?

Just to expand on the previous discussion: Reduce ops fall into two categories. "Pure" reduction ops apply an associative, commutative, binary operation to a set of values, and for these we use the identity-element for the binary operation (if it exists).  Examples are Sum, Product, Min, Max. The other reduction ops are composite-ops that follow a reduction-operation with other non-reduction operations. For these, the value is determined using the composite-ops to the identity-element of the reduction-operation. Examples are Mean, LogSum, etc.

**Comment by gramalingam** [comment]
@yuanyao-nv : do you have any further comments? This PR requires an approval from operator-sig and we would like to get this into the upcoming release (code-freeze tomorrow). Thanks!

**Review by yuanyao-nv — APPROVED** [review]
